### PR TITLE
Add Context.execute_with_session(s) functions

### PR DESCRIPTION
**Pull request by puiterwijk**
These are convenience functions to help with executing sections of code (Rust closures)
with a specific set of sessions.
This can avoid a lot of manual handling of sessions, and helps keep the session
use of each function clear (like shown in the PCR test case).

Signed-off-by: Patrick Uiterwijk <patrick@puiterwijk.org>

### CODE REVIEW

File: tests/context_tests.rs [hunk 1]
```diff
@@ -366,28 +366,26 @@ mod test_pcr_extend_reset {
...
-        context.clear_sessions();
         // Read PCR contents
         let pcr_selection_list = PcrSelectionListBuilder::new()
             .with_selection(HashingAlgorithm::Sha1, &[PcrSlot::Slot16])
             .with_selection(HashingAlgorithm::Sha256, &[PcrSlot::Slot16])
             .build();
-        let (_, _, pcr_data) = context.pcr_read(&pcr_selection_list).unwrap();
+        // pcr_read is NO_SESSIONS
+        let (_, _, pcr_data) =
+            context.execute_with_session(None, |ctx| ctx.pcr_read(&pcr_selection_list).unwrap());
```
Comment (ionut-arm) [inline tests/context_tests.rs:381]: Given that some calls are `NO_SESSIONS`, does it make sense to have an `execute_with_no_sessions` method as well?

File: tests/context_tests.rs [hunk 2]
```diff
@@ -366,28 +366,26 @@ mod test_pcr_extend_reset {
         // get reset by any D-RTPM events.
         // PCR (TCG PC Client Platform TPM Profile (PTP) for TPM 2.0 Version 1.05 Rev 14)
         let mut context = create_ctx_with_session();
-        let pcr_ses = context.sessions();
+        let pcr_ses = context.sessions().0;
 
         // We start by resetting. We do not place any expectations on the prior contents
-        context.pcr_reset(PcrHandle::Pcr16).unwrap();
+        context.execute_with_session(pcr_ses, |ctx| ctx.pcr_reset(PcrHandle::Pcr16).unwrap());
```
Comment (ionut-arm) [inline tests/context_tests.rs:372]: Looks like you're only using this `pcr_ses` session now, wouldn't it be easier to simply set it as the main session on `Context` and use the new methods only where you need a different one?

**Review by ionut-arm — COMMENTED** [review]
Thanks for the change 👌

File: tests/context_tests.rs (hunk 2 above)
Comment (puiterwijk) [reply tests/context_tests.rs:372]: I personally prefer to use the `execute_with_session` everywhere, to make it explicit which sessions are used with which calls.

File: tests/context_tests.rs (hunk 1 above)
Comment (puiterwijk) [reply tests/context_tests.rs:381]: Sure, I had exactly that in my own copy, will put that in here too.

File: tests/context_tests.rs (hunk 2 above)
Comment (puiterwijk) [reply tests/context_tests.rs:372]: If you prefer to use the `pcr_ses` by default, and only `execute_with_session` in the special case, I'll happily change. This is just my personal preference of explicit over implicit.

Comment (ionut-arm) [reply tests/context_tests.rs:372]: I don't mind either way, it's only a test! It does help with checking that the thing works properly 😄 So leave as is!

File: src/context.rs [hunk 3]
```diff
@@ -194,6 +194,45 @@ impl Context {
         self.sessions
     }
 
+    // Execute the closure in f with the specified set of sessions, and sets the original sessions back afterwards
+    // This is a convenience function
+    pub fn execute_with_sessions<F, T>(
```
Comment (ionut-arm) [inline src/context.rs:199]: Could you make this proper Rustdoc comments?

**Review by ionut-arm — APPROVED** [review]
Looks good to me, just one comment below

**Comment by Superhepper** [comment]
Nice! looks goof to me. They feel a lot like it would have come from the Rust std which is a good thing!

**Review by Superhepper — APPROVED** [review]

**Pull request merged**